src/messages/mod.rs

```diff
@@ -37,62 +35,114 @@ use xor_name::{Prefix, XorName};
 #[derive(Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
 pub struct Message {
     /// Destination location.
-    pub dst: DstLocation,
+    dst: DstLocation,
     /// Source authority.
-    pub src: SrcAuthority,
+    /// Messages do not need to sign this field as it is all verifiable (i.e. if the sig validates
+    /// agains the pub key and we know the pub key then we are good. If the proof is not recodnised we
+    /// ask for a longer chain that can be recodnised). Therefor we don't need to sign this field.
+    src: SrcAuthority,
     /// The body of the message.
-    pub variant: Variant,
+    variant: Variant,
     /// Source's knowledge of the destination section key. If present, the destination can use it
     /// to determine the length of the proof of messages sent to the source so the source would
     /// trust it (the proof needs to start at this key).
-    pub dst_key: Option<bls::PublicKey>,
-}
-
-/// Partially deserialized message.
-#[derive(Clone, Eq, PartialEq, Hash, Serialize, Deserialize, Debug)]
-pub struct PartialMessage {
-    /// Destination location.
-    pub dst: DstLocation,
-}
+    dst_key: Option<bls::PublicKey>,
 
-impl PartialMessage {
-    /// Deserialize the message.
-    pub fn from_bytes(bytes: &Bytes) -> Result<Self> {
-        Ok(bincode::deserialize(&bytes[..])?)
-    }
+    /// Serialised message, this is a signed and fully serialised message ready to send.
+    #[serde(skip)]
+    serialized: Bytes,
+    #[serde(skip)]
+    hash: MessageHash,
 }
 
 impl Message {
```

Comment:
Maybe we should add a method like this:
```rust
fn into_parts(self) -> (SrcAuthority, DstLocation, Option<bls::PublicKey>, Variant)
```
to avoid the `clone`s?

Reply:
I am not so sure, perhaps we can discuss this? I mean I am not seeing how that would work in practice and allowing the same functionality. Unless that method cloned each element itself. I may be missing the obvious though.

Reply:
Note it takes `self` by value so it can move from it:
```rust
fn into_parts(self) -> (SrcAuthority, DstLocation, Option<bls::PublicKey>, Variant) {
    (self.src, self.dst, self.dst_key, self.variant)
}
```

Reply:
Does that invalidate the message though when we move data from it?

Reply:
No, it consumes it, so it can't be used afterwards. It's just how moving works in rust (unlike C++ where you can still used a value after it has been moved from).

Reply:
Yes that's what I mean, So if you read msg.src() then msg.dst() the latter would fail unless all your data was in that parts tuple. But then you want to get the hash or serliazed msg are you not in trouble? I mean is it not more complex to take bits out like this? Even if we read msg.src more than once then we are back to clone() one of the calls. 
I am perhaps not understanding the value. Maybe it will take a bit of investigating where we clone() do we ever read it again? But then just making src dst return self.src etc. would achieve the same?

Reply:
My goal is the msg is immutable all the way through our code and we never have to call serialize or hash after construction/reception of the Message. I suppose if we did a `to_parts` we would need to be sure we don't any longer want to serialize or hash, maybe that's when the message is being handled? Although we would need to take care with bounced where we pass the received bytes back.

Reply:
> So if you read msg.src() then msg.dst() the latter would fail

No, you can't read `msg` after you called `into_parts` (or any other function that takes it by value). Rust statically prevents that. 

>  But then you want to get the hash or serliazed msg are you not in trouble

We can return these in the tuple as well if we think we might need them. But I think once we are in `dispatch_message` we no longer do.

> I am perhaps not understanding the value

It's just to avoid the `clone`. But maybe we don't have to worry about that just yet?

Reply:
Ah yes, I get you now, thanks. If you feel it's OK I would prefer to delay that decision until we finalise the message module proper. So in a later PR if at all?